planner/core/optimizer.go

```diff
@@ -386,6 +387,106 @@ func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
 	return plan
 }
 
+// prunePhysicalColumns currently only work for MPP(HashJoin<-Exchange).
+// Here add projection instead of pruning columns directly for safety considerations.
+// And projection is cheap here for it saves the network cost and work in memory.
+func prunePhysicalColumns(sctx sessionctx.Context, plan PhysicalPlan) {
+	emptyColumns := make([]*expression.Column, 0)
+	if tableReader, ok := plan.(*PhysicalTableReader); ok {
+		if _, isExchangeSender := tableReader.tablePlan.(*PhysicalExchangeSender); isExchangeSender {
+			prunePhysicalColumnsInternal(sctx, tableReader.tablePlan, emptyColumns, false)
+		}
```

Comment:
If the child is not `PhysicalExchangeSender`, do we need to call `prunePhysicalColumns` recursively for each child?